src/scheduler/Scheduler.cpp

```diff
@@ -1129,4 +1164,191 @@ ExecGraphNode Scheduler::getFunctionExecGraphNode(unsigned int messageId)
 
     return node;
 }
+
+void Scheduler::checkForMigrationOpportunities()
+{
+    std::vector<std::shared_ptr<faabric::PendingMigrations>>
+      tmpPendingMigrations;
+
+    {
+        // Acquire a shared lock to read from the in-flight requests map
+        faabric::util::SharedLock lock(mx);
+
+        tmpPendingMigrations = doCheckForMigrationOpportunities();
+    }
+
+    // If we find migration opportunites
+    if (tmpPendingMigrations.size() > 0) {
+        // Acquire full lock to write to the pending migrations map
+        faabric::util::FullLock lock(mx);
+
+        for (auto msgPtr : tmpPendingMigrations) {
+            // First, broadcast the pending migrations to other hosts
+            broadcastAddPendingMigration(msgPtr);
+            // Second, update our local records
+            pendingMigrations[msgPtr->appid()] = std::move(msgPtr);
+        }
+    }
+}
+
+void Scheduler::broadcastAddPendingMigration(
+  std::shared_ptr<faabric::PendingMigrations> pendingMigrations)
+{
+    // Get all hosts for the to-be migrated app
+    auto msg = pendingMigrations->migrations().at(0).msg();
+    std::string funcStr = faabric::util::funcToString(msg, false);
+    std::set<std::string>& thisRegisteredHosts = registeredHosts[funcStr];
+
+    // Remove this host from the set
+    registeredHosts.erase(thisHost);
+
+    // Send pending migrations to all involved hosts
+    for (auto& otherHost : thisRegisteredHosts) {
+        getFunctionCallClient(otherHost).sendAddPendingMigration(
+          pendingMigrations);
+    }
+}
+
+std::shared_ptr<faabric::PendingMigrations> Scheduler::canAppBeMigrated(
+  uint32_t appId)
+{
+    faabric::util::SharedLock lock(mx);
+
+    if (pendingMigrations.find(appId) == pendingMigrations.end()) {
+        return nullptr;
+    }
+
+    return pendingMigrations[appId];
+}
+
+void Scheduler::addPendingMigration(
+  std::shared_ptr<faabric::PendingMigrations> pMigration)
+{
+    faabric::util::FullLock lock(mx);
+
+    auto msg = pMigration->migrations().at(0).msg();
+    if (pendingMigrations.find(msg.appid()) != pendingMigrations.end()) {
+        SPDLOG_ERROR("Received remote request to add a pending migration for "
+                     "app {}, but already recorded another migration request"
+                     " for the same app.",
+                     msg.appid());
+        throw std::runtime_error("Remote request for app already there");
+    }
+
+    pendingMigrations[msg.appid()] = pMigration;
+}
+
+void Scheduler::removePendingMigration(uint32_t appId)
+{
+    faabric::util::FullLock lock(mx);
+
+    inFlightRequests.erase(appId);
+    pendingMigrations.erase(appId);
+}
+
+std::vector<std::shared_ptr<faabric::PendingMigrations>>
+Scheduler::doCheckForMigrationOpportunities(
```

Comment:
The need for a new method other than re-using `makeSchedulingDecisions`  arises from the fact that:
* In this method we just check for the ability to migrate, but do not actually change the underlying host resources.
* The scheduling and migration policies can differ. The former being more complex and the latter simpler.